### DISCUSSION TITLE
The auth().onAuthStateChanged() is NOT initially triggered for SIGNED OUT users

### DISCUSSION THREAD

**Discussion by cbdeveloper** [invertase/react-native-firebase#8148] [Q&A] [answered]
Once my app loads, I have this `useEffect` to init user authentication.

```
import auth, { firebase, FirebaseAuthTypes } from '@react-native-firebase/auth';

(...)

useEffect(() => {
  console.log('Running useEffect...');
  console.log('auth().currentUser', auth().currentUser?.email);
  const unsubscribe = auth().onAuthStateChanged((firebaseUser) => {
    console.log('Running onAuthStateChanged...');
    console.log('firebaseUser', firebaseUser?.email);
    setUser(firebaseUser);
    setIsInitializing(false);
    hideSplash();
  });
  return unsubscribe;
}, []);
```

This is what happens when user is already signed-in:

```
LOG  Running useEffect...
LOG  auth().currentUser: xxx@example.com
LOG  Running onAuthStateChanged...       // <---- onAuthStateChanged is triggered
LOG  firebaseUser: xxx@example.com
```

This is what happens when user is initially signed out:

```
LOG  Running useEffect...
LOG  auth().currentUser null
// <---- onAuthStateChanged is NOT triggered
```

When I realized that it does get triggered in the `SIGNED_IN` scenerio, but not in the `SIGNED_OUT` scenario, I thought this was weird. Is this the intended behavior?

Version:

```
"@react-native-firebase/app": "^20.4.0",
"@react-native-firebase/auth": "^20.4.0",
"@react-native-firebase/firestore": "^20.4.0",
"@react-native-firebase/functions": "^20.4.0",
"expo": "~51.0.28",
"react-native": "0.74.5",
```

**Comment by mikehardy** [comment, marked as answer]
Can you update and retest? Posting issues or discussions on old versions is unproductive - I'm certain this area just changed recently because I changed it, to fix an issue that sounds pretty similar to this...

https://github.com/invertase/react-native-firebase/blob/main/CHANGELOG.md#2150-2024-11-16
https://github.com/invertase/react-native-firebase/pull/8139
https://github.com/invertase/react-native-firebase/issues/7819

**Reply by cbdeveloper** [reply]
Hey Mike, sorry about that. I've updated to `v21.6.0` and here are the results.

It seems like there is a difference between first load and the subsequent loads of the app while using the `expo-dev-client`. After further testing, I've found that this is probably related to how Expo dev client handles the app reloads. I was testing on a local `dev` build. 

Now that I've tested in `preview` build (without the Expo dev client), it looks like it's being triggered just fine. If I kill and reopen the app, I can always see the splash being hidden (which happens inside the `onAuthStateChanged` handler), regardless if user is signed in or out.

Thanks a lot for checking in.

```
Android Bundled 4315ms node_modules/expo/AppEntry.js (2568 modules)             // Loaded the app
 LOG  Running useEffect...              // Set onAuthStateChanged listener
 LOG  auth().currentUser null
 LOG  Running onAuthStateChanged...     // On first load it seems to trigger it for signed out
 LOG  firebaseUser null
 LOG  Running onAuthStateChanged...     // Then I've signed in and it triggers
 LOG  firebaseUser xxx@example.com
 LOG  Rendering Paywall...
Android Bundled 149ms node_modules/expo/AppEntry.js (1 module)                  // Reloaded the app
 LOG  Running useEffect...              // Set onAuthStateChanged listener
 LOG  auth().currentUser xxx@example.com
 LOG  Running onAuthStateChanged...     // It triggers for the signed in case
 LOG  firebaseUser xxx@example.com
 LOG  Rendering Paywall...
 LOG  Running onAuthStateChanged...     // Then I've signed out and it triggers
 LOG  firebaseUser null
Android Bundled 149ms node_modules/expo/AppEntry.js (1 module)                  // Reloaded the app
 LOG  Running useEffect...             // Set onAuthStateChanged listener
 LOG  auth().currentUser null
                                       // It does not trigger for the signed out case
```

**Reply by mikehardy** [reply]
Interesting. I'd say that's still an error, the call trace is very useful, I should be able to reproduce that.
What platform are you reproducing this on - android, or iOS (or both?)

**Reply by cbdeveloper** [reply]
Hey Mike, 

I don't remember exactly what I was using to test that day. I think it was on my Android device.

So today I tried reproducing it both on iOS simulator and Android device, but it didn't happen again.

Today, both platforms were triggering the `onAuthStateChanged` listener for `null` user both for initial loads and app reloads using the Expo dev client.

Maybe it was some weird caching issue on the dev client. If I run into it again, I'll update this thread with the specific details.

Again, thanks for the amazing support!

---

**iOS Simulator** | iPhone SE (3rd gen) iOS 16.4

```
iOS Bundled 4233ms node_modules/expo/AppEntry.js (2550 modules) // Initial load for null user
 LOG  Running useEffect...
 LOG  auth().currentUser null
 LOG  Running onAuthStateChanged...
 LOG  firebaseUser null
iOS Bundled 130ms node_modules/expo/AppEntry.js (1 module) // Reloaded the dev client
 LOG  Running useEffect...
 LOG  auth().currentUser null
 LOG  Running onAuthStateChanged...
 LOG  firebaseUser null
 LOG  Running onAuthStateChanged...  // Signed in
 LOG  firebaseUser xxx@example.com
 LOG  Running onAuthStateChanged... // Signed out
 LOG  firebaseUser null
iOS Bundled 125ms node_modules/expo/AppEntry.js (1 module) // Reloaded dev client
 LOG  Running useEffect...
 LOG  auth().currentUser null
 LOG  Running onAuthStateChanged...
 LOG  firebaseUser null
iOS Bundled 137ms node_modules/expo/AppEntry.js (1 module) // Reloaded dev client
 LOG  Running useEffect...
 LOG  auth().currentUser null
 LOG  Running onAuthStateChanged...
 LOG  firebaseUser null
```

Galaxy S21 device (Android 14)

```
Android Bundled 4037ms node_modules/expo/AppEntry.js (2558 modules) // Initial load for null user
 LOG  Running useEffect...
 LOG  auth().currentUser null
 LOG  Running onAuthStateChanged...
 LOG  firebaseUser null
Android Bundled 112ms node_modules/expo/AppEntry.js (1 module) // Reloaded dev client
 LOG  Running useEffect...
 LOG  auth().currentUser null
 LOG  Running onAuthStateChanged...
 LOG  firebaseUser null
Android Bundled 106ms node_modules/expo/AppEntry.js (1 module) // Reloaded dev client
 LOG  Running useEffect...
 LOG  auth().currentUser null
 LOG  Running onAuthStateChanged...
 LOG  firebaseUser null
 LOG  Running onAuthStateChanged... // Signed in
 LOG  firebaseUser xxx@example.com
 LOG  Running onAuthStateChanged... // Signed out
 LOG  firebaseUser null
Android Bundled 112ms node_modules/expo/AppEntry.js (1 module) // Reloaded dev client
 LOG  Running useEffect...
 LOG  auth().currentUser null
 LOG  Running onAuthStateChanged...
 LOG  firebaseUser null
```

**Reply by mikehardy** [reply]
Okay, I've marked my "please update and retest" as the answer since the updated code appears to be working
This exact problem was definitely occurring, it was a real problem but I think it's settled now, glad it appears so for you
Good luck with your project!
